File: uva/524 Prime Ring/ring.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

vector<int> primes = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31};

int ans[16] = {1};

void recurse(int n, int used, int remaining, int prev) {
    if (remaining == 0) {
        if (binary_search(primes.begin(), primes.end(), prev + 1)) {
            for (int i = 0; i < n - 1; i ++) {
                cout << ans[i] << " ";
            }
            cout << ans[n - 1] << "\n";
        }
    }
    for (int i = 1; i <= n; i ++) {
        if (!(used & (2 << (i - 1)))) {
            if (binary_search(primes.begin(), primes.end(), (prev + i))) {
                ans[n - remaining] = i;
                recurse(n, used | (2 << (i - 1)), remaining - 1, i);
            }
        }
    }
}


int main () {
    int s;
    int i = 1;
    while (cin >> s) {
        if (i > 1) {
            cout << endl;
        }
        cout << "Case " << i++ << ":\n";
        recurse(s, 2, s - 1, 1);
    }
}

File: uva/524 Prime Ring/ring.py
primes = set([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31])



def recurse(n, used, remaining, prev):
    if remaining == 0:
        if (prev + 1) in primes:
            print(*ans)
        else:
            return
    for i in range(1, n + 1):
        if not (used & (2 << (i - 1))):
            if (prev + i) in primes:
                ans[n - remaining] = i
                recurse(n, used | (2 << (i - 1)), remaining - 1, i)
    return ans

s = int(input())
i = 0;
while True:
    ans = [0] * s
    ans[0] = 1
    i += 1
    print("Case {}:".format(i))
    #for l in recurse(s, 2, s - 1, 1):
        #if (l[-1] + 1) in primes:
        #print(*([1] + l))
    recurse(s, 2, s - 1, 1)
    try:
        s = int(input())
        print("")
    except:
        exit(0)